### PR TITLE
FR-10132 - split cookie and type fixes

### PR DESCRIPTION
**Pull request by yuvalotem1**
Split cookie fixes
- make split cookie work correctly 
- adjust remove cookies to new structure
- create CookieManager class and refactor all cookie logic to be there

Type fixes
- remove TS error when using withFronteggApp on next 13

Tests
- Tests for cookie manager
- Tests for utils

### CODE REVIEW

File: packages/nextjs/src/common/helpers.ts [hunk 1]
```diff
@@ -80,24 +80,29 @@ function createSplitCookie(cookieName: string, session: string, options: CookieS
 function chunkString(str: string, numChunks: number) {
```
Comment (raz-shlomo-frontegg) [inline packages/nextjs/src/common/helpers.ts:80]: add typings  return type`: string[]`

File: packages/nextjs/src/common/helpers.ts [hunk 2]
```diff
@@ -107,9 +112,45 @@ export function addToCookies(newCookies: string[], res: ServerResponse) {
...
+const createEmptySingleCookie = (cookieName: string, isSecured: boolean, cookieDomain: string) =>
+  createCookie({ cookieName, session: '', expires: new Date(), isSecured, cookieDomain });
+
+const createEmptyCookies = (cookieName: string, isSecured: boolean, cookieDomain: string, numberOfCookies: number) => {
+  if (!numberOfCookies || numberOfCookies === 1) {
+    return createEmptySingleCookie(cookieName, isSecured, cookieDomain);
+  }
+  const allEmptyCookies = [];
+  for (let i = 0; i < numberOfCookies; i++) {
+    allEmptyCookies.push(...createEmptySingleCookie(`${cookieName}-${i}`, isSecured, cookieDomain));
```
Comment (raz-shlomo-frontegg) [inline packages/nextjs/src/common/helpers.ts:124]: missing +1 in the cookie name

File: packages/nextjs/src/common/helpers.ts [hunk 3]
```diff
@@ -107,9 +112,45 @@ export function addToCookies(newCookies: string[], res: ServerResponse) {
   }
   res.setHeader('set-cookie', [...existingSetCookie, ...newCookies]);
 }
+const createEmptySingleCookie = (cookieName: string, isSecured: boolean, cookieDomain: string) =>
+  createCookie({ cookieName, session: '', expires: new Date(), isSecured, cookieDomain });
+
+const createEmptyCookies = (cookieName: string, isSecured: boolean, cookieDomain: string, numberOfCookies: number) => {
```
Comment (raz-shlomo-frontegg) [inline packages/nextjs/src/common/helpers.ts:118]: i suggest to create splitcookie manager class

File: packages/nextjs/src/common/helpers.ts [hunk 4]
```diff
@@ -107,9 +112,45 @@ export function addToCookies(newCookies: string[], res: ServerResponse) {
...
+  }
+  return allEmptyCookies;
+};
+
+export const getNumberOfCookiesFromRequest = (req?: RequestType): number => {
+  let numberOfCookies = 1;
+  if (!req) {
+    return numberOfCookies;
+  }
+  const cookieStr = 'credentials' in req ? req.headers.get('cookie') || '' : req.headers.cookie || '';
```
Comment (raz-shlomo-frontegg) [inline packages/nextjs/src/common/helpers.ts:134]: consider extract or document

File: packages/nextjs/src/common/helpers.ts [hunk 5]
```diff
@@ -107,9 +112,45 @@ export function addToCookies(newCookies: string[], res: ServerResponse) {
...
+  const allCookies = cookieStr && cookie.parse(cookieStr);
```
Comment (raz-shlomo-frontegg) [inline packages/nextjs/src/common/helpers.ts:135]: verify that cookie parse doesn't throw. if do, catch it

File: packages/nextjs/src/common/helpers.ts [hunk 6]
```diff
@@ -107,9 +112,45 @@ export function addToCookies(newCookies: string[], res: ServerResponse) {
...
+  if (!allCookies || allCookies[FronteggConfig.cookieName]) {
+    return numberOfCookies;
+  }
+  while (allCookies[`${FronteggConfig.cookieName}-${numberOfCookies}`]) {
+    numberOfCookies++;
+  }
+  return numberOfCookies - 1;
```
Comment (raz-shlomo-frontegg) [inline packages/nextjs/src/common/helpers.ts:142]: consider change the cookie structure to be zero-based (start with 0)

File: packages/nextjs/src/common/helpers.ts [hunk 7]
```diff
@@ -107,9 +112,45 @@ export function addToCookies(newCookies: string[], res: ServerResponse) {
...
-export function removeCookies(cookieName: string, isSecured: boolean, cookieDomain: string, res: ServerResponse) {
-  const cookieValue = createCookie({ cookieName, session: '', expires: new Date(), isSecured, cookieDomain });
+export function removeCookies(
+  cookieName: string,
+  isSecured: boolean,
+  cookieDomain: string,
+  res: ServerResponse,
+  req?: RequestType
+) {
+  const numberOfCookies = getNumberOfCookiesFromRequest(req);
```
Comment (raz-shlomo-frontegg) [inline packages/nextjs/src/common/helpers.ts:152]: consider change to remove by name and not by amount

File: packages/nextjs/src/withFronteggApp.tsx [hunk 8]
```diff
@@ -12,7 +12,7 @@ type FronteggCustomApp = NextComponentType<
   AppPropsType
 >;
 export const withFronteggApp = (
-  app: ((props: AppProps) => JSX.Element) & {
+  app: ((props: any) => JSX.Element) & {
```
Comment (raz-shlomo-frontegg) [inline packages/nextjs/src/withFronteggApp.tsx:15]: replace with `unknown`

File: packages/nextjs/src/common/helpers.ts (hunk 6 above)
Comment (yuvalotem1) [reply packages/nextjs/src/common/helpers.ts:142]: will keep like this for now

**Review by raz-shlomo-frontegg — APPROVED** [review]

**Comment by raz-shlomo-frontegg** [comment]
I would add tests.
I am really suggest to having tests for this section

**Review by raz-shlomo-frontegg — APPROVED** [review]

**Review by raz-shlomo-frontegg — APPROVED** [review]

**Pull request merged**